storage/remote/queue_manager.go

```diff
@@ -688,40 +732,43 @@ func (s *shards) start(n int) {
 	s.qm.metrics.numShards.Set(float64(n))
 
 	newQueues := make([]chan sample, n)
+	buffer := make([]chan []sample, n)
 	for i := 0; i < n; i++ {
 		newQueues[i] = make(chan sample, s.qm.cfg.Capacity)
+		buffer[i] = make(chan []sample, s.qm.cfg.Capacity)
 	}
-
+	s.queueBuf = buffer
 	s.queues = newQueues
+	s.firstSendPending = true
 
 	var hardShutdownCtx context.Context
 	hardShutdownCtx, s.hardShutdown = context.WithCancel(context.Background())
 	s.softShutdown = make(chan struct{})
 	s.running.Store(int32(n))
+	s.reshard = make(chan struct{})
 	s.done = make(chan struct{})
 	s.droppedOnHardShutdown.Store(0)
 	for i := 0; i < n; i++ {
 		go s.runShard(hardShutdownCtx, i, newQueues[i])
 	}
 }
 
-// stop the shards; subsequent call to enqueue will return false.
-func (s *shards) stop() {
-	// Attempt a clean shutdown, but only wait flushDeadline for all the shards
-	// to cleanly exit.  As we're doing RPCs, enqueue can block indefinitely.
-	// We must be able so call stop concurrently, hence we can only take the
-	// RLock here.
+// stop existing shards and triggers flushing of samples that are existing in the queue.
+func (s *shards) stop(resharding bool) {
 	s.mtx.RLock()
 	close(s.softShutdown)
 	s.mtx.RUnlock()
-
-	// Enqueue should now be unblocked, so we can take the write lock.  This
-	// also ensures we don't race with writes to the queues, and get a panic:
-	// send on closed channel.
 	s.mtx.Lock()
 	defer s.mtx.Unlock()
-	for _, queue := range s.queues {
-		close(queue)
+	for i := range s.queues {
+		close(s.queues[i])
+	}
+	if resharding {
+		s.qm.blockIncomingSamples.Store(true)
+		return
+	}
```

Comment:
I don't really like that this resharding logic is done inside of stop. A shard stop should just worry about itself, not the queue manager.

Reply:
Yeah, this was wrong. I have shifted this to `reshardLoop()` which controls the transfer of shards, etc. This should be the place for blocking.